### DISCUSSION TITLE
Overall features overview and questions

### DISCUSSION THREAD

**Discussion by thednp** [vanjs-org/van#323] [Q&A] [answered]
### Hi @Tao-VanJS!

Man, I love this! To say this is just another UI framework... I stumbled upon this a couple days ago and read everything I could find about it. If I could put all my knowledge into a fresh new code repository it would be something like VanJS, which is indeed one of the fastest, most ergonomic (really write less code to same effect), probably one of the best out there. As one of the guys showcasing the script puts it in his video: it's not as elegant as Solid, but it definitely gets the job done.

However I have a couple of question, probably most of them already addressed, but here we go:

### Router
The community solutions I found `vanjs-router` and `vanjs-routing` seem fairly basic, which is fine, but IMO most apps need a combination of both, and a little bit more. While having to write my own routes is a preferred choice instead of having to use some sort of Regex based FileSystemRouting, I would also consider a way to split the code for obvious reasons. While VanJS is 1Kb, a real app can be 1500Kb+ of code which needs to be split. So the question is: are there any plans for improvement here?

### Server Actions and Server Functions
I love what @RyanSolid did with SolidStart, especially in regards to Server Actions and Server Functions. If we combine them with a little better router, I think the result is going to be of the best UX/DX these days. As explained above, I'm thinking of a way to lazy load routes and deliver HTML without the double data problem. Any plans in this direction?

### Examples
I think we need more examples with these metaframework like features, I can try and help with that here.

I think this might be a real turning point for VanJS, to learn from others and try to do better, to make a difference, it's a great opportunity for innovation. All that costs time, energy and effort, and I don't know what your take is in these regards, just I feel it would be a shame not to have your work put on the map.

Thank you Tao. Love your work regardless.
DNP

**Comment by Tao-VanJS** [comment, marked as answer]
Hi @thednp,

Thank you so much for your detailed feedback and comments!

I will address the points you mentioned in this post.

First and foremost, as you can see. VanJS tries to organize itself in a "microkernel" architecture, with a minimalist core library, while allowing other meta-frameworks implemented on top of the core to extend its functionalities. Thus for the features that you wish to have but haven't been there yet, a new add-on could be a good way to go.

Regarding the points you mentioned for router, my view is that with the gradual adoption of `URLPattern` (and the eventual standardization in the future), it will become a much easier problem to solve, potentially to a point where you can have very maintainable and ergonomic code without even any framework. On the other hand, as an unopinionated framework, VanJS doesn't want to force its users to split their code in a specific way. Users should make the final decision on how to split different components into individual files. Users should be able to choose from combining all the components in a single file (for a small project), or split them into modules based on a specific rule and import them to use.

Regarding server functions and server actions, my view is that they can be understood as syntax sugars which hide the details of sending and processing HTTP requests. I understand that some people prefer this omission and some IDE benefit for easier jumping into definition. On the other hand, in principle, VanJS wants to be transpiling-free. I believe this is crucial for its simplicity as this principle will give users the confidence that what they see in the code is exactly what's being executed.

That said, even though these features are at odds with the principles of VanJS's core. As I mentioned above, with a "microkernel" architecture, you're free to build add-ons to make the developer experience of VanJS the way you like. For instance, even though I personally believe JSX adds the complexity of UI framework thus increases the entry barrier, I understand there are people like this style of syntax thus the [van-jsx](https://github.com/vanjs-org/van/blob/main/addons/van_jsx) add-on is helpful for these people.

Thanks again for your comments!

**Reply by thednp** [reply]
I hope we can find some time to do a joint work creating some better examples with SSR (nitro backend) and something like @tanstack/router. That would be awesome!
